Vagrantfile

```diff
@@ -1,71 +1,302 @@
+  define_ubuntu config, name: 'precise64', dist: 'precise', bits: 64, llvm: { url: "http://crystal-lang.s3.amazonaws.com/llvm/llvm-3.9.1-1-linux-x86_64.tar.gz", path: "llvm-3.9.1-1" }
+  define_ubuntu config, name: 'precise32', dist: 'precise', bits: 32, llvm: { url: "http://crystal-lang.s3.amazonaws.com/llvm/llvm-3.9.1-1-linux-i686.tar.gz", path: "llvm-3.9.1-1" }
+
+  define_debian config, name: 'stretch64', dist: 'stretch', bits: 64
+  define_debian config, name: 'jessie64', dist: 'jessie', bits: 64
+
+  define_freebsd config, name: 'freebsd11', box: '11.2-STABLE'
+
+  define_alpine config, name: 'alpine64', bits: 64
+
+  config.vm.provider "virtualbox" do |vb|
+    vb.memory = 6*1024
+    vb.cpus = 2
+
+    # Keep time synced with host
+    # vb.customize [ "guestproperty", "set", :id, "/VirtualBox/GuestAdd/VBoxService/--timesync-set-threshold", 10000 ]
+  end
+end
+
+def clone_crystal_from_vagrant(config)
+  # use ~/crystal directory instead of /vagrant
+  # to have a clean copy of working directory
+  # in the native fs of the vm
   config.vm.provision :shell, privileged: false, inline: %(
     git clone /vagrant crystal
   )
 end
 
-Vagrant.configure("2") do |config|
-  %w(precise trusty xenial).product([32, 64]).each do |dist, bits|
-    box_name = "#{dist}#{bits}"
+def setup_memory(config, bits)
+  if bits == 32
+    config.vm.provider "virtualbox" do |vb|
+      vb.memory = 4*1024
+    end
+  end
+end
+
+# *llvm* values: 6.0, 7, (empty), Hash with {url:, path:}
+def install_llvm(dist, llvm)
+  llvm ||= "6.0"
+
+  if llvm.is_a?(Hash)
+    %(
+      curl -s #{llvm[:url]} | tar xz -C /opt
+      echo 'export PATH="$PATH:/opt/#{llvm[:path]}/bin"' >> /etc/profile.d/crystal.sh
+    )
+  else
+    llvm_suffix = llvm != "" ? "-#{llvm}" : ""
+
+    %(
+      add-apt-repository "deb http://apt.llvm.org/#{dist}/ llvm-toolchain-#{dist}#{llvm_suffix} main"
+      curl -sSL https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add -
+      apt-get update
+      apt-get install -y llvm#{llvm_suffix}-dev
+    )
+  end
+end
+
+# instructions to install crystal binary on all platforms
+def install_crystal(c, family, dist:, bits:, github_targz:)
+  if github_targz
+    fail "github .tar.gz installation method for #{family} is not supported" unless family == "ubuntu" || family == "debian"
 
-    config.vm.define(box_name) do |c|
-      c.vm.box = "ubuntu/#{box_name}"
+    targz_url = bits == 64 ? CRYSTAL_LINUX64_TARGZ : CRYSTAL_LINUX32_TARGZ
+
+    c.vm.provision :shell, inline: %(
+      mkdir -p /opt/crystal
+      echo '#{targz_url}'
+      curl -sSL --output - '#{targz_url}' | tar xz -C /opt/crystal --strip-component=1 -f -
+
+      echo 'export LIBRARY_PATH=/opt/crystal/lib/crystal/lib/:${LIBRARY_PATH}' >> /etc/profile.d/crystal.sh
+      echo 'export PATH=/opt/crystal/bin:$PATH' >> /etc/profile.d/crystal.sh
+    )
+  else
+    case family
+    when "ubuntu"
```

Comment:
Why not merge the instructions for Ubuntu and Debian? They're essentially the same.

Reply:
Easier to follow I guess. I would leave them as duplicated actually.